File: go/internal/httpapi/replaylab.go
// Replay Lab routes (T-517; reference run-routes/replay-lab.ts): the
// whole-run sandbox replay (optionally with a patched workflow) and the
// targeted fork. Editor + runs.start, "ai" rate family, org-scoped
// source with identical 404s (no enumeration leak), nested labs
// rejected, and the reference's exact error-code ladder.
package httpapi

import (
	"encoding/json"
	"net/http"
	"time"

	"github.com/johnny4young/janusly/go/internal/aiconfig"
	"github.com/johnny4young/janusly/go/internal/audit"
	"github.com/johnny4young/janusly/go/internal/auth"
	"github.com/johnny4young/janusly/go/internal/domain"
	"github.com/johnny4young/janusly/go/internal/ratelimit"
	"github.com/johnny4young/janusly/go/internal/store"
)

const replayLabOverrideMaxBytes = 64_000

// resolveLabSnapshot loads the source run's workflow: saved
// workflow_versions.dagJson first, ad-hoc runs.inputJson.workflow as the
// fallback. Also returns the source's trigger-time input for propagation.
func (s *V1Server) resolveLabSnapshot(r *http.Request, rc v1Request, run store.GetRunRow) (*domain.Workflow, any, *opResult) {
	q := store.New(s.pool)
	var snapshot []byte
	if version, err := q.GetWorkflowVersionByID(r.Context(), store.GetWorkflowVersionByIDParams{
		ID: run.WorkflowVersionID, OrgID: rc.orgID,
	}); err == nil && len(version.DagJson) > 0 {
		snapshot = version.DagJson
	}
	var envelope struct {
		Workflow json.RawMessage `json:"workflow"`
		Input    any             `json:"input"`
	}
	_ = json.Unmarshal(run.InputJson, &envelope)
	if len(snapshot) == 0 {
		snapshot = envelope.Workflow
	}
	if len(snapshot) == 0 || string(snapshot) == "null" {
		failed := opError(http.StatusBadRequest, "no_workflow_snapshot",
			"Source run has no workflow snapshot available", nil)
		return nil, nil, &failed
	}
	wf, issues := domain.Parse(snapshot)
	if wf == nil {
		reason := "unknown"
		if len(issues) > 0 {
			reason = issues[0].Message
		}
		failed := opError(http.StatusBadRequest, "invalid_snapshot",
			"Source run snapshot failed schema validation: {{reason}}", map[string]any{"reason": reason})
		return nil, nil, &failed
	}
	return wf, envelope.Input, nil
}

// labSource org-scopes the source run and rejects nested sandboxes.
func (s *V1Server) labSource(r *http.Request, rc v1Request, sourceRunID, notFoundCode string) (store.GetRunRow, *opResult) {
	run, err := store.New(s.pool).GetRun(r.Context(), store.GetRunParams{ID: sourceRunID, OrgID: rc.orgID})
	if err != nil {
		failed := opError(http.StatusNotFound, notFoundCode, "Source run not found", nil)
		return store.GetRunRow{}, &failed
	}
	if run.ReplayMode.Valid && run.ReplayMode.String != "" {
		failed := opError(http.StatusBadRequest, "nested_replay_lab",
			"Source run is itself a sandbox run; cannot start a nested lab", nil)
		return store.GetRunRow{}, &failed
	}
	return run, nil
}

func (s *V1Server) labRateLimit(r *http.Request, rc v1Request) *opResult {
	_, settings := aiconfig.Resolve(r.Context(), s.pool, rc.orgID)
	if err := s.limiter.Enforce(r.Context(), rc.orgID, ratelimit.Options{
		Name: "ai", Max: settings.RateLimitPerMin, Window: time.Minute,
	}); err != nil {
		failed := opError(http.StatusTooManyRequests, "rate_limited", err.Error(), nil)
		return &failed
	}
	return nil
}

func (s *V1Server) replayLabCore(r *http.Request, rc v1Request) opResult {
	if rejection := s.labRateLimit(r, rc); rejection != nil {
		return *rejection
	}
	var body struct {
		SourceRunID       string          `json:"sourceRunId"`
		SuggestedWorkflow json.RawMessage `json:"suggestedWorkflow"`
	}
	if err := decodeBody(r, &body); err != nil || body.SourceRunID == "" {
		return opError(http.StatusBadRequest, "runs_source_run_id_required", "sourceRunId is required", nil)
	}
	run, rejection := s.labSource(r, rc, body.SourceRunID, "runs_source_run_not_found")
	if rejection != nil {
		return *rejection
	}

	hasPatch := len(body.SuggestedWorkflow) > 0 && string(body.SuggestedWorkflow) != "null"
	var wf *domain.Workflow
	var triggerInput any
	if hasPatch {
		var probe any
		if json.Unmarshal(body.SuggestedWorkflow, &probe) != nil {
			return opError(http.StatusBadRequest, "runs_suggested_workflow_not_object", "suggestedWorkflow must be an object", nil)
		}
		if _, ok := probe.(map[string]any); !ok {
			return opError(http.StatusBadRequest, "runs_suggested_workflow_not_object", "suggestedWorkflow must be an object", nil)
		}
		parsed, issues := domain.Parse(body.SuggestedWorkflow)
		if parsed == nil {
			reason := "unknown"
			if len(issues) > 0 {
				reason = issues[0].Message
			}
			return opError(http.StatusBadRequest, "runs_suggested_workflow_invalid",
				"suggestedWorkflow failed schema validation: {{reason}}", map[string]any{"reason": reason})
		}
		if blocking := validateGeneratedWorkflow(body.SuggestedWorkflow); len(blocking) > 0 {
			return opError(http.StatusBadRequest, "runs_suggested_workflow_sanitize_failed",
				"suggestedWorkflow sanitize failed: {{reason}}", map[string]any{"reason": blocking[0].Message})
		}
		wf = parsed
		var envelope struct {
			Input any `json:"input"`
		}
		_ = json.Unmarshal(run.InputJson, &envelope)
		triggerInput = envelope.Input
	} else {
		var failed *opResult
		wf, triggerInput, failed = s.resolveLabSnapshot(r, rc, run)
		if failed != nil {
			return *failed
		}
	}

	replayRunID, err := s.engine.ReplayRunAsValidation(r.Context(), rc.orgID, body.SourceRunID, wf, triggerInput, rc.userID, hasPatch)
	if err != nil {
		return opError(http.StatusInternalServerError, "internal_error", "Internal error: "+err.Error(), nil)
	}
	audit.Write(r.Context(), s.pool, rc.authContext, "replay_lab.started", audit.Options{
		TargetType: "run", TargetID: body.SourceRunID,
		Metadata: map[string]any{"replayRunId": replayRunID, "hasPatch": hasPatch},
	})
	return opOK(map[string]any{"runId": replayRunID})
}

func (s *V1Server) replayLabForkCore(r *http.Request, rc v1Request) opResult {
	if rejection := s.labRateLimit(r, rc); rejection != nil {
		return *rejection
	}
	var body struct {
		SourceRunID   string          `json:"sourceRunId"`
		ForkNodeID    string          `json:"forkNodeId"`
		InputOverride json.RawMessage `json:"inputOverride"`
	}
	if err := decodeBody(r, &body); err != nil || body.SourceRunID == "" {
		return opError(http.StatusBadRequest, "invalid_input", "sourceRunId is required", nil)
	}
	if body.ForkNodeID == "" {
		return opError(http.StatusBadRequest, "invalid_input", "forkNodeId is required", nil)
	}
	var override any
	if len(body.InputOverride) > 0 && string(body.InputOverride) != "null" {
		if len(body.InputOverride) > replayLabOverrideMaxBytes {
			return opError(http.StatusUnprocessableEntity, "override_too_large",
				"inputOverride exceeds 64 KiB cap ({{overrideBytes}} bytes)",
				map[string]any{"overrideBytes": len(body.InputOverride)})
		}
		if json.Unmarshal(body.InputOverride, &override) != nil {
			return opError(http.StatusBadRequest, "invalid_input", "inputOverride must be valid JSON", nil)
		}
	}
	run, rejection := s.labSource(r, rc, body.SourceRunID, "not_found")
	if rejection != nil {
		return *rejection
	}
	// v1 forks replay the source's OWN snapshot (patches go through the
	// whole-run lab so validation covers the full DAG).
	wf, triggerInput, failed := s.resolveLabSnapshot(r, rc, run)
	if failed != nil {
		return *failed
	}
	result, err := s.engine.ReplayRunAsValidationFork(r.Context(), rc.orgID, body.SourceRunID, wf, body.ForkNodeID, override, triggerInput, rc.userID)
	if err != nil {
		return opError(http.StatusInternalServerError, "internal_error", "Internal error: "+err.Error(), nil)
	}
	if !result.OK {
		return opError(http.StatusUnprocessableEntity, result.Code, result.Message, nil)
	}
	audit.Write(r.Context(), s.pool, rc.authContext, "replay_lab.fork_started", audit.Options{
		TargetType: "run", TargetID: body.SourceRunID,
		Metadata: map[string]any{
			"replayRunId": result.RunID, "forkNodeId": body.ForkNodeID,
			"predecessorCount": result.PredecessorCount, "hasOverride": override != nil,
		},
	})
	return opOK(map[string]any{"runId": result.RunID, "predecessorCount": result.PredecessorCount})
}

func (s *V1Server) mountReplayLabRoutes(mux *http.ServeMux) {
	s.route(mux, "POST /runs/replay-lab", routeGate{auth.RoleEditor, "runs.start"}, func(w http.ResponseWriter, r *http.Request, rc v1Request) {
		writeLegacy(w, s.replayLabCore(r, rc))
	})
	s.route(mux, "POST /runs/replay-lab/fork", routeGate{auth.RoleEditor, "runs.start"}, func(w http.ResponseWriter, r *http.Request, rc v1Request) {
		writeLegacy(w, s.replayLabForkCore(r, rc))
	})
}

File: apps/api/src/routes/run-routes/replay-lab.ts
/** Whole-run and targeted Replay Lab routes. */

import { and, eq } from "drizzle-orm";

import { db, runs, workflowVersions } from "@janusly/db";
import {
  replayRunAsValidation,
  replayRunAsValidationFork,
} from "@janusly/engine/src/adapters/replay-lab";
import { WorkflowSchema, type Workflow } from "@janusly/shared";

import { MAX_JSON_BODY_BYTES } from "../../api-config";
import { auditAction } from "../../audit-helper";
import { orgLlmRuntime, sanitizeAiWorkflow } from "../../ai-runtime";
import { RATE_LIMIT_WINDOW_MS } from "../../constants";
import { asRecord, readJson, sendError, sendJson } from "../../http";
import { enforceRateLimit } from "../../rate-limit";
import type { Route } from "../../routes";

export const replayLabRoutes: Route[] = [
  // Replay Lab — standalone sandbox replay. Creates a fresh validation
  // run from ANY source run (not just a DLQ entry) and re-executes the
  // workflow from root nodes. The source run's workflow snapshot is
  // pulled from `workflow_versions.dagJson` (saved workflow) or from
  // `runs.inputJson.workflow` (ad-hoc fallback); an optional caller
  // `suggestedWorkflow` overrides the snapshot and is validated through
  // the same `WorkflowSchema` + `sanitizeAiWorkflow` chain `/dlq/validate-fix`
  // uses. The new run carries `replayMode = "validation"` so the engine's
  // dryRun gating, write-side skips, and rollup exclusions apply
  // automatically; the audit row distinguishes the lab intent from the
  // recovery-dialog intent (`replay_lab.started` vs
  // `recovery.validation_started`).
  { method: "POST", match: "/runs/replay-lab", role: "editor", permission: "runs.start",
    handler: async ({ req, res, auth }) => {
      const { orgConfig } = await orgLlmRuntime(auth.orgId);
      await enforceRateLimit(auth.orgId, { name: "ai", windowMs: RATE_LIMIT_WINDOW_MS, max: orgConfig.ai.rateLimitPerMin });
      const body = asRecord(await readJson(req, MAX_JSON_BODY_BYTES));

      const sourceRunId = typeof body.sourceRunId === "string" ? body.sourceRunId : null;
      if (!sourceRunId) return sendError(res, "runs_source_run_id_required", "sourceRunId is required", 400);

      // Org-scope the source run and reject cross-org / unknown ids with
      // an identical 404 envelope — no enumeration leak.
      const sourceRows = await db.select().from(runs).where(and(eq(runs.id, sourceRunId), eq(runs.orgId, auth.orgId)));
      const sourceRun = sourceRows[0];
      if (!sourceRun) return sendError(res, "runs_source_run_not_found", "Source run not found", 404);

      // No nested labs — a sandbox run is itself a validation run; replaying
      // it would just create a sibling sandbox with the same snapshot.
      if (sourceRun.replayMode) {
        return sendError(res, "nested_replay_lab", "Source run is itself a sandbox run; cannot start a nested lab", 400);
      }

      // Resolve the workflow snapshot. Precedence: caller-supplied patch >
      // saved workflow_versions.dagJson > ad-hoc runs.inputJson.workflow.
      let workflow: Workflow;
      const hasPatch = body.suggestedWorkflow !== undefined && body.suggestedWorkflow !== null;
      if (hasPatch) {
        if (typeof body.suggestedWorkflow !== "object") {
          return sendError(res, "runs_suggested_workflow_not_object", "suggestedWorkflow must be an object", 400);
        }
        const parsed = WorkflowSchema.safeParse(body.suggestedWorkflow);
        if (!parsed.success) {
          return sendError(
            res,
            "runs_suggested_workflow_invalid",
            "suggestedWorkflow failed schema validation: {{reason}}",
            400,
            { reason: parsed.error.issues[0]?.message ?? "unknown" },
          );
        }
        try {
          workflow = sanitizeAiWorkflow(parsed.data);
        } catch (err) {
          const reason = err instanceof Error ? err.message : String(err);
          return sendError(res, "runs_suggested_workflow_sanitize_failed", "suggestedWorkflow sanitize failed: {{reason}}", 400, { reason });
        }
      } else {
        // No patch — load the source's own workflow snapshot. Try the saved
        // version first (most common), fall back to the ad-hoc inputJson.
        const versionRows = await db
          .select()
          .from(workflowVersions)
          .where(and(eq(workflowVersions.id, sourceRun.workflowVersionId), eq(workflowVersions.orgId, auth.orgId)));
        const version = versionRows[0];
        let snapshot: unknown = version?.dagJson;
        if (!snapshot) {
          const input = sourceRun.inputJson as Record<string, unknown> | null;
          snapshot = input && typeof input === "object" ? input.workflow : null;
        }
        if (!snapshot) {
          return sendError(res, "no_workflow_snapshot", "Source run has no workflow snapshot available; supply suggestedWorkflow", 400);
        }
        const parsed = WorkflowSchema.safeParse(snapshot);
        if (!parsed.success) {
          return sendError(
            res,
            "invalid_snapshot",
            "Source run snapshot failed schema validation: {{reason}}",
            400,
            { reason: parsed.error.issues[0]?.message ?? "unknown" },
          );
        }
        workflow = parsed.data;
      }

      // Propagate trigger-time input from the source run so workflows
      // that reference `{{input.*}}` behave identically in the sandbox.
      // `startRun` writes `inputJson: { workflow, input }` — read the
      // same key here. Falls back to `{}` for ad-hoc runs whose
      // inputJson predates the input field.
      const sourceInputJson = sourceRun.inputJson as Record<string, unknown> | null;
      const triggerInput = sourceInputJson && typeof sourceInputJson === "object"
        ? sourceInputJson.input
        : undefined;

      const { runId: replayRunId } = await replayRunAsValidation({
        orgId: auth.orgId,
        sourceRunId,
        workflow,
        input: triggerInput,
        createdBy: auth.userId,
        hasPatch,
      });

      await auditAction(auth, "replay_lab.started", { targetType: "run", targetId: sourceRunId, metadata: {
        replayRunId,
        hasPatch,
      } });

      return sendJson(res, { runId: replayRunId });
    } },
  // Targeted Replay Lab fork — re-run a workflow starting at one node
  // instead of from scratch. Predecessors of the fork node are cloned
  // from the source run's terminal state (status='succeeded' with
  // stateJson copied) so the fork node can read upstream outputs without
  // re-paying the cost of HTTP / AI / tool calls that already succeeded.
  // Same `replayMode='validation'` flag as the whole-run replay-lab so
  // the engine's dryRun gate skips write-side effects uniformly.
  // Audit action `replay_lab.fork_started` distinguishes forks from
  // whole-run replays at compliance read time.
  { method: "POST", match: "/runs/replay-lab/fork", role: "editor", permission: "runs.start",
    handler: async ({ req, res, auth }) => {
      const { orgConfig } = await orgLlmRuntime(auth.orgId);
      await enforceRateLimit(auth.orgId, { name: "ai", windowMs: RATE_LIMIT_WINDOW_MS, max: orgConfig.ai.rateLimitPerMin });
      const body = asRecord(await readJson(req, MAX_JSON_BODY_BYTES));

      const sourceRunId = typeof body.sourceRunId === "string" ? body.sourceRunId : null;
      if (!sourceRunId) return sendError(res, "invalid_input", "sourceRunId is required", 400);

      const forkNodeId = typeof body.forkNodeId === "string" ? body.forkNodeId : null;
      if (!forkNodeId) return sendError(res, "invalid_input", "forkNodeId is required", 400);

      // Defensive cap on inputOverride size — same shape as the audit
      // metadata cap. Over-cap landings would balloon `run_nodes.stateJson`
      // and `run_events.payload` later; reject up-front with a clear code.
      if (body.inputOverride !== undefined) {
        const overrideBytes = Buffer.byteLength(JSON.stringify(body.inputOverride), "utf8");
        if (overrideBytes > 64_000) {
          return sendError(
            res,
            "override_too_large",
            "inputOverride exceeds 64 KiB cap ({{overrideBytes}} bytes)",
            422,
            { overrideBytes },
          );
        }
      }

      // Org-scope the source run. Cross-org / unknown id → 404 (no enumeration leak).
      const sourceRows = await db.select().from(runs).where(and(eq(runs.id, sourceRunId), eq(runs.orgId, auth.orgId)));
      const sourceRun = sourceRows[0];
      if (!sourceRun) return sendError(res, "not_found", "Source run not found", 404);

      // No nested forks — source run can't itself be a sandbox run.
      if (sourceRun.replayMode) {
        return sendError(res, "nested_replay_lab", "Source run is itself a sandbox run; cannot fork from it", 400);
      }

      // Resolve the workflow snapshot. v1 forks DO NOT accept a patched
      // workflow — they replay the source's own snapshot at the fork node
      // (patches go through the whole-run lab path so the patch validation
      // covers the full DAG). If a future use case wants "fork + patch",
      // it's a separate route.
      const versionRows = await db
        .select()
        .from(workflowVersions)
        .where(and(eq(workflowVersions.id, sourceRun.workflowVersionId), eq(workflowVersions.orgId, auth.orgId)));
      const version = versionRows[0];
      let snapshot: unknown = version?.dagJson;
      if (!snapshot) {
        const input = sourceRun.inputJson as Record<string, unknown> | null;
        snapshot = input && typeof input === "object" ? input.workflow : null;
      }
      if (!snapshot) {
        return sendError(res, "no_workflow_snapshot", "Source run has no workflow snapshot available; cannot fork", 400);
      }
      const parsed = WorkflowSchema.safeParse(snapshot);
      if (!parsed.success) {
        return sendError(
          res,
          "invalid_snapshot",
          "Source run snapshot failed schema validation: {{reason}}",
          400,
          { reason: parsed.error.issues[0]?.message ?? "unknown" },
        );
      }
      const workflow = parsed.data;

      // Propagate trigger-time input so `{{input.*}}` references resolve
      // to the same values the source run saw at predecessors.
      const sourceInputJson = sourceRun.inputJson as Record<string, unknown> | null;
      const triggerInput = sourceInputJson && typeof sourceInputJson === "object"
        ? sourceInputJson.input
        : undefined;

      const result = await replayRunAsValidationFork({
        orgId: auth.orgId,
        sourceRunId,
        workflow,
        forkNodeId,
        inputOverride: body.inputOverride,
        input: triggerInput,
        createdBy: auth.userId,
      });

      if (!result.ok) {
        // Map adapter discriminated errors to HTTP codes. `fork_node_not_found`
        // is a 422 (caller supplied a value the route accepted shape-wise
        // but the workflow doesn't have that node). `predecessor_not_succeeded`
        // is a 422 (we can't fork past an unreliable upstream).
        return sendError(res, result.code, result.message, 422);
      }

      await auditAction(auth, "replay_lab.fork_started", { targetType: "run", targetId: sourceRunId, metadata: {
        replayRunId: result.runId,
        forkNodeId,
        predecessorCount: result.predecessorCount,
        hasOverride: body.inputOverride !== undefined,
      } });

      return sendJson(res, {
        runId: result.runId,
        predecessorCount: result.predecessorCount,
      });
    } },
];
